Vamos a utilizar un base de datos de cancer de mama y lo importaremos con sklearn. Usaremos (as_frame=True) para que nos pueda hacer un dataframe

In [3]:
from sklearn import datasets

datos = datasets.load_breast_cancer(as_frame=True)

Vemos los datos

In [5]:
datos.frame.head()

,mean radius,mean texture,mean perimeter,mean area,mean smoothness,mean compactness,mean concavity,mean concave points,mean symmetry,mean fractal dimension,...,worst texture,worst perimeter,worst area,worst smoothness,worst compactness,worst concavity,worst concave points,worst symmetry,worst fractal dimension,target
0,17.99,10.38,122.80,1001.0,0.11840,0.27760,0.3001,0.14710,0.2419,0.07871,...,17.33,184.60,2019.0,0.1622,0.6656,0.7119,0.2654,0.4601,0.11890,0
1,20.57,17.77,132.90,1326.0,0.08474,0.07864,0.0869,0.07017,0.1812,0.05667,...,23.41,158.80,1956.0,0.1238,0.1866,0.2416,0.1860,0.2750,0.08902,0
2,19.69,21.25,130.00,1203.0,0.10960,0.15990,0.1974,0.12790,0.2069,0.05999,...,25.53,152.50,1709.0,0.1444,0.4245,0.4504,0.2430,0.3613,0.08758,0
3,11.42,20.38,77.58,386.1,0.14250,0.28390,0.2414,0.10520,0.2597,0.09744,...,26.50,98.87,567.7,0.2098,0.8663,0.6869,0.2575,0.6638,0.17300,0
4,20.29,14.34,135.10,1297.0,0.10030,0.13280,0.1980,0.10430,0.1809,0.05883,...,16.67,152.20,1575.0,0.1374,0.2050,0.4000,0.1625,0.2364,0.07678,0


Tambien podemos usar esto para ver mas información

In [6]:
datos.frame.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 569 entries, 0 to 568
Data columns (total 31 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   mean radius              569 non-null    float64
 1   mean texture             569 non-null    float64
 2   mean perimeter           569 non-null    float64
 3   mean area                569 non-null    float64
 4   mean smoothness          569 non-null    float64
 5   mean compactness         569 non-null    float64
 6   mean concavity           569 non-null    float64
 7   mean concave points      569 non-null    float64
 8   mean symmetry            569 non-null    float64
 9   mean fractal dimension   569 non-null    float64
 10  radius error             569 non-null    float64
 11  texture error            569 non-null    float64
 12  perimeter error          569 non-null    float64
 13  area error               569 non-null    float64
 14  smoothness error         5

Y como podemos ver no hay datos vacios

Y lo podemos confirmar

In [7]:
datos.frame.isna(). sum()

,0
mean radius,0
mean texture,0
mean perimeter,0
mean area,0
mean smoothness,0
mean compactness,0
mean concavity,0
mean concave points,0
mean symmetry,0
mean fractal dimension,0


O podemos usar (datos.frame.describe()) para obtener los datos de mejor manera

In [8]:
datos.frame.describe()

,mean radius,mean texture,mean perimeter,mean area,mean smoothness,mean compactness,mean concavity,mean concave points,mean symmetry,mean fractal dimension,...,worst texture,worst perimeter,worst area,worst smoothness,worst compactness,worst concavity,worst concave points,worst symmetry,worst fractal dimension,target
count,569.000000,569.000000,569.000000,569.000000,569.000000,569.000000,569.000000,569.000000,569.000000,569.000000,...,569.000000,569.000000,569.000000,569.000000,569.000000,569.000000,569.000000,569.000000,569.000000,569.000000
mean,14.127292,19.289649,91.969033,654.889104,0.096360,0.104341,0.088799,0.048919,0.181162,0.062798,...,25.677223,107.261213,880.583128,0.132369,0.254265,0.272188,0.114606,0.290076,0.083946,0.627417
std,3.524049,4.301036,24.298981,351.914129,0.014064,0.052813,0.079720,0.038803,0.027414,0.007060,...,6.146258,33.602542,569.356993,0.022832,0.157336,0.208624,0.065732,0.061867,0.018061,0.483918
min,6.981000,9.710000,43.790000,143.500000,0.052630,0.019380,0.000000,0.000000,0.106000,0.049960,...,12.020000,50.410000,185.200000,0.071170,0.027290,0.000000,0.000000,0.156500,0.055040,0.000000
25%,11.700000,16.170000,75.170000,420.300000,0.086370,0.064920,0.029560,0.020310,0.161900,0.057700,...,21.080000,84.110000,515.300000,0.116600,0.147200,0.114500,0.064930,0.250400,0.071460,0.000000
50%,13.370000,18.840000,86.240000,551.100000,0.095870,0.092630,0.061540,0.033500,0.179200,0.061540,...,25.410000,97.660000,686.500000,0.131300,0.211900,0.226700,0.099930,0.282200,0.080040,1.000000
75%,15.780000,21.800000,104.100000,782.700000,0.105300,0.130400,0.130700,0.074000,0.195700,0.066120,...,29.720000,125.400000,1084.000000,0.146000,0.339100,0.382900,0.161400,0.317900,0.092080,1.000000
max,28.110000,39.280000,188.500000,2501.000000,0.163400,0.345400,0.426800,0.201200,0.304000,0.097440,...,49.540000,251.200000,4254.000000,0.222600,1.058000,1.252000,0.291000,0.663800,0.207500,1.000000


Usaremos datos.data.shape para ver la forma que tienen los datos

In [9]:
datos.data.shape

(569, 30)

Son 569 registros con 30 caracteristicas

Ahora crearemos el modelo para hacer las predicciones

In [11]:
from sklearn.model_selection import train_test_split

X = datos.data
y = datos.target

X_ent, X_pru, y_ent, y_pru = train_test_split(X, y, test_size= .3)

In [13]:
from sklearn import svm
from sklearn import metrics

modelo = svm.SVC()
modelo.fit(X_ent, y_ent)
predicciones = modelo.predict(X_pru)
print("Exactitud:", metrics.accuracy_score(y_pru, predicciones))
print("Precisión:", metrics.precision_score(y_pru, predicciones))

Exactitud: 0.9064327485380117
Precisión: 0.8728813559322034


Haremos una lista de los diferentes kernels que podemos usar y tambien podemos usar diferentes valores de gamma

In [16]:
from sklearn import svm
from sklearn import metrics

kernels = ["linear", "rbf", "sigmoid"] # Changed 'Linear' to 'linear'
gamma = [1, .01, .001, .0001, .00001]

for kernel in kernels:
  for gamma_val in gamma:
    modelo = svm.SVC(kernel=kernel, gamma=gamma_val)
    modelo.fit(X_ent, y_ent)
    predicciones = modelo.predict(X_pru)
    print(f"{kernel}: Gamma: {gamma_val}: Exactitud:", metrics.accuracy_score(y_pru, predicciones))
    print(f"{kernel}: Gamma: {gamma_val}: Precisión:", metrics.precision_score(y_pru, predicciones))

linear: Gamma: 1: Exactitud: 0.935672514619883
linear: Gamma: 1: Precisión: 0.9043478260869565
linear: Gamma: 0.01: Exactitud: 0.935672514619883
linear: Gamma: 0.01: Precisión: 0.9043478260869565
linear: Gamma: 0.001: Exactitud: 0.935672514619883
linear: Gamma: 0.001: Precisión: 0.9043478260869565
linear: Gamma: 0.0001: Exactitud: 0.935672514619883
linear: Gamma: 0.0001: Precisión: 0.9043478260869565
linear: Gamma: 1e-05: Exactitud: 0.935672514619883
linear: Gamma: 1e-05: Precisión: 0.9043478260869565
rbf: Gamma: 1: Exactitud: 0.6081871345029239
rbf: Gamma: 1: Precisión: 0.6081871345029239
rbf: Gamma: 0.01: Exactitud: 0.6081871345029239
rbf: Gamma: 0.01: Precisión: 0.6081871345029239
rbf: Gamma: 0.001: Exactitud: 0.9239766081871345
rbf: Gamma: 0.001: Precisión: 0.9174311926605505
rbf: Gamma: 0.0001: Exactitud: 0.9239766081871345
rbf: Gamma: 0.0001: Precisión: 0.9099099099099099
rbf: Gamma: 1e-05: Exactitud: 0.9064327485380117
rbf: Gamma: 1e-05: Precisión: 0.8793103448275862
sigmoid: Ga

Ahora especificaremos que el kernel sea linear

In [18]:
modelo = svm.SVC(kernel="linear")
modelo.fit(X_ent, y_ent)
predicciones = modelo.predict(X_pru)
print(f"{kernel}: Gamma: {gamma_val}: Exactitud:", metrics.accuracy_score(y_pru, predicciones))
print(f"{kernel}: Gamma: {gamma_val}: Precisión:", metrics.precision_score(y_pru, predicciones))

sigmoid: Gamma: 1e-05: Exactitud: 0.935672514619883
sigmoid: Gamma: 1e-05: Precisión: 0.9043478260869565


Y hacemos el reporte de calificaciones

In [19]:
from sklearn.metrics import classification_report

print(classification_report(y_pru, predicciones))

              precision    recall  f1-score   support

           0       1.00      0.84      0.91        67
           1       0.90      1.00      0.95       104

    accuracy                           0.94       171
   macro avg       0.95      0.92      0.93       171
weighted avg       0.94      0.94      0.93       171



Y ya por ultimo lo metemos en un DataFrame

In [20]:
from sklearn.metrics import confusion_matrix
import pandas as pd

pd.DataFrame(confusion_matrix(y_pru, predicciones))

,0,1
0,56,11
1,0,104
